pkg/proc/bininfo.go

```diff
@@ -427,44 +435,128 @@ func (bi *BinaryInfo) PCToFunc(pc uint64) *Function {
 	}
 	for _, image := range bi.Images {
 		if image.Path == path && image.addr == addr {
-			return
+			return true
 		}
 	}
-	//TODO(aarzilli): actually load informations about the image here
-	bi.Images = append(bi.Images, &Image{Path: path, addr: addr})
+	return false
+}
+
+// AddImage adds the specified image to bi, loading data asynchronously.
+// Addr is the relocated entry point for the executable and staticBase (i.e.
+// the relocation offset) for all other images.
+// The first image added must be the executable file.
+func (bi *BinaryInfo) AddImage(path string, addr uint64) error {
+	if bi.hasImage(path, addr) {
+		return nil
+	}
+	image := &Image{Path: path, addr: addr, typeCache: make(map[dwarf.Offset]godwarf.Type)}
+	// add Image regardless of error so that we don't attempt to re-add it every time we stop
+	image.index = len(bi.Images)
+	bi.Images = append(bi.Images, image)
+	err := loadBinaryInfo(bi, image, path, addr)
+	if err != nil {
+		bi.Images[len(bi.Images)-1].loadErr = err
+	}
+	return err
+}
+
+// moduleDataToImage finds the image corresponding to the given module data object.
+func (bi *BinaryInfo) moduleDataToImage(md *moduleData) *Image {
+	return bi.funcToImage(bi.PCToFunc(uint64(md.text)))
+}
+
+// imageToModuleData finds the module data in mds corresponding to the given image.
+func (bi *BinaryInfo) imageToModuleData(image *Image, mds []moduleData) *moduleData {
+	for _, md := range mds {
+		im2 := bi.moduleDataToImage(&md)
+		if im2.index == image.index {
+			return &md
+		}
+	}
+	return nil
+}
+
+// typeToImage returns the image containing the give type.
+func (bi *BinaryInfo) typeToImage(typ godwarf.Type) *Image {
+	return bi.Images[typ.Common().Index]
 }
 
 // Close closes all internal readers.
 func (bi *BinaryInfo) Close() error {
-	if bi.sepDebugCloser != nil {
-		bi.sepDebugCloser.Close()
+	var errs []error
+	for _, image := range bi.Images {
+		if image.sepDebugCloser != nil {
+			err := image.sepDebugCloser.Close()
+			if err != nil {
+				errs = append(errs, fmt.Errorf("closing shared object %q (split dwarf): %v", image.Path, err))
+			}
+		}
+		if image.closer != nil {
+			err := image.closer.Close()
+			if err != nil {
+				errs = append(errs, fmt.Errorf("closing shared object %q: %v", image.Path, err))
+			}
+		}
 	}
-	if bi.closer != nil {
-		return bi.closer.Close()
+	switch len(errs) {
+	case 0:
+		return nil
+	case 1:
+		return errs[0]
+	default:
+		return fmt.Errorf("multiple errors closing executable files")
 	}
-	return nil
 }
 
-func (bi *BinaryInfo) setLoadError(fmtstr string, args ...interface{}) {
-	bi.loadErrMu.Lock()
-	bi.loadErr = fmt.Errorf(fmtstr, args...)
-	bi.loadErrMu.Unlock()
+func (image *Image) setLoadError(fmtstr string, args ...interface{}) {
+	image.loadErrMu.Lock()
+	image.loadErr = fmt.Errorf(fmtstr, args...)
+	image.loadErrMu.Unlock()
 }
 
-// LoadError returns any internal load error.
+// LoadError returns any internal error incurred while loading the last shared object.
```

Comment:
Why return just the last error specifically?

Reply:
IIRC it was the easiest way to do it at the time, because it was called from the backends. Given that now it's only called in proc.PostInitializationSetup, I've replaced it there with a loop through all the images.